### ISSUE TITLE
Program crashes after upgraing to rpy2 v3.5.2

### ISSUE THREAD

**Issue by mohyddintash** [rpy2/rpy2#888]
**Describe the issue or bug**

I just upgraded to 3.5.2 and my program crashes when calling r functions in my r package from python and passing in a string value.
here's the console output:

```bash
File "/root/.local/share/virtualenvs/src-NVTF7jWz/lib/python3.9/site-packages/rpy2/robjects/functions.py", line 202, in __call__

return (super(SignatureTranslatedFunction, self)

File "/root/.local/share/virtualenvs/src-NVTF7jWz/lib/python3.9/site-packages/rpy2/robjects/functions.py", line 117, in __call__

new_args = [cv.py2rpy(a) for a in args]

File "/root/.local/share/virtualenvs/src-NVTF7jWz/lib/python3.9/site-packages/rpy2/robjects/functions.py", line 117, in <listcomp>

  new_args = [cv.py2rpy(a) for a in args]

File "/usr/lib/python3.9/functools.py", line 877, in wrapper

  return dispatch(args[0].__class__)(*args, **kw)

File "/root/.local/share/virtualenvs/src-NVTF7jWz/lib/python3.9/site-packages/rpy2/robjects/conversion.py", line 240, in _py2rpy

  raise NotImplementedError(

NotImplementedError: Conversion 'py2rpy' not defined for objects of type '<class 'str'>'
```

**To Reproduce**

Steps to reproduce the behavior are generally required.
call an r function with string input ex: 

`result = client.my_r_package.my_r_function(my_string)`

**Expected behavior**
Program runs and the r function is executed successfully.
**Error**
NotImplementedError: Conversion 'py2rpy' not defined for objects of type '<class 'str'>'

**Additional context**
This works just fine in 3.5.1

**Comment by lgautier** [comment]
Can you share self-contained minimal example to reproduce the issue?

**Comment by adkinsrs** [comment]
I can also reproduce this issue, but the NotImplementedError changes depending on what line the error occurs on in the stacktrace

```python3
import sys  # for print debugging
import traceback    # for debugging

import rpy2.robjects as ro
from rpy2.robjects.packages import importr
from rpy2.robjects import pandas2ri
from rpy2.robjects.conversion import localconverter
from rpy2.robjects.vectors import ListVector, StrVector
from rpy2.rinterface_lib import openrlib

class RError(Exception):
    """Error based on issues that would manifest in any particular R-language call."""
    def __init__(self, message="") -> None:
        self.message = message
        super().__init__(self.message)

def convert_r_df_to_r_matrix(df):
    """
    Convert pandas dataframe to R-style matrix
    """
    r_matrix = ro.r["as.matrix"]
    return r_matrix(df)

def convert_r_matrix_to_r_df(mtx):
    """
    Convert R-style matrix to R-style dataframe
    """
    # mtx is a matrix of numbers with PCs in columns
    r_df = ro.r["as.data.frame"]
    return r_df(mtx)

def convert_r_matrix_to_r_prcomp(mtx):
    """
    Convert R-style matrix to R-style prcomp object
    """
    # mtx is a matrix of numbers with PCs in columns
    prcomp_obj = ListVector({"rotation": mtx})
    prcomp_obj.rclass = "prcomp"    # Convert to prcomp-class object
    return prcomp_obj

def run_projectR_cmd(target_df, loading_df, is_pca=False):
    """
    Convert input Pandas dataframes to R matrix.
    Pass the inputs into the projectR function written in R.
    Return Pandas dataframe of the projectR output
    """

    # R does not play nice with multithreading so a lock is necessary to prevent interruptions
    try:
        projectR = importr('projectR')
    except Exception as e:
        print("ERROR: {}".format(str(e)), file=sys.stderr)
        print(traceback.print_exc(), file=sys.stderr)
        raise RError("Could not import projectR package.")

    # Convert from pandas dataframe to R data.frame
    with localconverter(ro.default_converter + pandas2ri.converter):
        target_r_df = ro.conversion.py2rpy(target_df)
        loading_r_df = ro.conversion.py2rpy(loading_df)

    # data.frame to matrix (projectR has no data.frame signature)
    target_r_matrix = convert_r_df_to_r_matrix(target_r_df)
    loading_r_matrix = convert_r_df_to_r_matrix(loading_r_df)

    # Assign Rownames to each matrix
    target_r_matrix.rownames = StrVector(target_df.index)
    loading_r_matrix.rownames = StrVector(loading_df.index)

    # Modify the R-style matrix to be a prcomp object if necessary
    loading_r_object = loading_r_matrix
    if is_pca:
        loading_r_object = convert_r_matrix_to_r_prcomp(loading_r_matrix)

    with openrlib.rlock:
        # Run project R command.  Get projectionPatterns matrix
        try:
            projection_patterns_r_matrix = projectR.projectR(data=target_r_matrix, loadings=loading_r_object, full=False)
        except Exception as e:
            print("ERROR: {}".format(str(e)), file=sys.stderr)
            print(traceback.print_exc(), file=sys.stderr)
            raise RError("Error: Could not run projectR command.")

    # matrix back to data.frame
    projection_patterns_r_df = convert_r_matrix_to_r_df(projection_patterns_r_matrix)

    # Convert from R data.frame to pandas dataframe
    with localconverter(ro.default_converter + pandas2ri.converter):
        projection_patterns_df = ro.conversion.rpy2py(projection_patterns_r_df)

    return projection_patterns_df
```

For reference, I call the `run_projectR_cmd` elsewhere , and the indexes in both `target_df` and `loading_df` need to overlap, and the values are numeric.

Typically the error occurs on the `projectR = importr('projectR')` line with the following error: `NotImplementedError: Conversion 'rpy2py' not defined for objects of type '<class 'rpy2.rinterface_lib.sexp.SexpEnvironment'>'`

If I move that line outside of the function (global), I now get the error occurring at the line `return r_matrix(df)` which is within the `convert_df_to_r_matrix` function.  This error, however, is for a different type `NotImplementedError: Conversion 'rpy2py' not defined for objects of type '<class 'rpy2.rinterface.SexpClosure'>'`

Seems that the conversion error shows based on the first line of R-code that rpy2 needs to interpret and execute.  I backversioned to 3.5.1, and the problem does not seem to occur there.  I am also running this in a Docker environment (ubuntu:impish base image), using python-3.10.4 and R-3.6.3

**Comment by lgautier** [comment]
If the error occurs when calling `importr()` a shorter example would be:

```python
import sys  # for print debugging
import traceback    # for debugging

import rpy2.robjects as ro
from rpy2.robjects.packages import importr
from rpy2.robjects import pandas2ri
from rpy2.robjects.conversion import localconverter
from rpy2.robjects.vectors import ListVector, StrVector
from rpy2.rinterface_lib import openrlib

projectR = importr('projectR')
```

I don't have Bioconductor installed, and the issue is not present with any other package I am trying. Could you point was is the specific objects in `projectR` that is causing the error?

**Comment by adkinsrs** [comment]
Hi @lgautier 

I'm not entirely sure if `projectR` is causing the issue, as if I got a similar error when calling `target_r_matrix = convert_r_df_to_r_matrix(target_r_df)`.

However I did downgrade rpy2 to v3.5.1, and this seemed to resolve any issues I was having.

**Comment by lgautier** [comment]
Can you work out a minimal example with `convert_r_df_to_r_matrix` then?

**Comment by adkinsrs** [comment]
Hi @lgautier. I will try to get you a minimal example by the end of next week. I have some big deadlines early next week that depend on the current working codebase on my project.  After that, I can replicate the condition with less code.

**Comment by adkinsrs** [comment]
@lgautier I have a minimal example ready.  For this I have a tab-delimited file as input.  For this, I am using rpy2 3.5.2 on python 3.9.  Typically I run my code with rpy2 in a Docker container but for this example I was able to replicate it in the REPL in a non-Docker environment.

```python3
import pandas as pd
df = pd.read_csv("./320ca057-0119-4f32-8397-7761ea084ed1.csv", delimiter="\t")    # tab-delimited CSV file

import rpy2.robjects as ro
def convert_r_df_to_r_matrix(df):
    """
    Convert pandas dataframe to R-style matrix
    """
    r_matrix = ro.r["as.matrix"]
    return r_matrix(df)

matrix = convert_r_df_to_r_matrix(df)

"""
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 6, in convert_r_df_to_r_matrix
  File "/opt/homebrew/lib/python3.9/site-packages/rpy2/robjects/functions.py", line 202, in __call__
    return (super(SignatureTranslatedFunction, self)
  File "/opt/homebrew/lib/python3.9/site-packages/rpy2/robjects/functions.py", line 117, in __call__
    new_args = [cv.py2rpy(a) for a in args]
  File "/opt/homebrew/lib/python3.9/site-packages/rpy2/robjects/functions.py", line 117, in <listcomp>
    new_args = [cv.py2rpy(a) for a in args]
  File "/opt/homebrew/Cellar/python@3.9/3.9.12/Frameworks/Python.framework/Versions/3.9/lib/python3.9/functools.py", line 888, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/opt/homebrew/lib/python3.9/site-packages/rpy2/robjects/conversion.py", line 240, in _py2rpy
    raise NotImplementedError(
NotImplementedError: Conversion 'py2rpy' not defined for objects of type '<class 'pandas.core.frame.DataFrame'>'
"""
```

It is worth noting that I do not have the "projectR" package installed on my non-Docker environment (more or less even Bioconductor).  I am attempting to install it now, and see if the error jumps around if I add the projectR import statement before or after the `convert_r_df_to_r_matrix` function call.

**Comment by adkinsrs** [comment]
@lgautier I actually updated my R version in docker from 3.6.3 to 4.2.0 and this seems to have resolved my particular weird errors in rpy2 v3.5.2.

In my non-Docker (M1 Mac) terminal I also ran the following in a python REPL and everything worked as intended.

```python
#python version - 3.9.13
#R version - 4.2.1
#rpy2 version. - 3.5.2

# Read in tab-delimited CSV file into dataframe (22836 rows x 144 columns)
import pandas as pd
df = pd.read_csv("./320ca057-0119-4f32-8397-7761ea084ed1.csv", delimiter="\t")

from rpy2.robjects import pandas2ri
from rpy2.robjects.conversion import localconverter
with localconverter(ro.default_converter + pandas2ri.converter):
    r_df = ro.conversion.py2rpy(df)

# High-level interface implementation ... works!
import rpy2.robjects as ro
def convert_r_df_to_r_matrix(df):
    """
    Convert pandas dataframe to R-style matrix
    """
    r_matrix = ro.r["as.matrix"]
    return r_matrix(df)

matrix = convert_r_df_to_r_matrix(r_df)

# Low-level interface implementation ... works!
import rpy2.rinterface as ri
def convert_r_df_to_r_matrix_low(df):
    r_matrix = ri.baseenv["as.matrix"]
    return r_matrix(df)
matrix_low = convert_r_df_to_r_matrix_low(r_df)
```

I was not able to add the `projectR = importr('projectR')` command to this snippet as I am having issues installing R packages on my M1 Mac due to some clang architecture argument things. So I think it is safe to say that a simple upgrade of R resolved the original issue.  😀

**Comment by lgautier** [comment]
Closing since upgrading R fixed things. ARM64 Mac (M1, M2) seems to be still tricky for some setups but hopefully getting better.
